Shiny_Stats_Server.r

```diff
@@ -0,0 +1,1212 @@
+    } else if(input$type_QQ == "Normal"){
+      rnorm(n)
+    } else if(input$type_QQ == "Negatively skewed"){
+      rbeta(n, 15, 2)
+    } else if(input$type_QQ == "Positively skewed"){
+      rgamma(n, 2)
+    }
+    
+    par(mfrow = c(2, 2), mar = c(2, 2, 2, 2), oma = c(0, 0, 5, 0))
+    
+    qqnorm(x)
+    qqline(x, col = "red")
+    ### Parameters of corresponding normal distribution
+    mu <- mean(x)
+    sigma <- sd(x)
+    ### Create vectors m quantiles of normal and chosen distributions
+    normal_quantiles <- qnorm(c(.001, seq(.05, .99, .05), .999), mu, sigma)
+    names(normal_quantiles) <- paste0(seq(0, 100, 5), "%")
+    sample_quantiles <- quantile(x, seq(0, 1, .05))
+    
+    if (input$type_QQ == "Light-tailed"){
+      density_quantiles <- dunif(sample_quantiles)
+    } else if(input$type_QQ == "Heavy-tailed"){
+      density_quantiles <- dt(sample_quantiles, 2)
+    } else if(input$type_QQ == "Normal"){
+      density_quantiles <- dnorm(sample_quantiles, mu, sigma)
+    } else if(input$type_QQ == "Negatively skewed"){
+      density_quantiles <- dbeta(sample_quantiles, 15, 2)
+    } else if(input$type_QQ == "Positively skewed"){
+      density_quantiles <- dgamma(sample_quantiles, 2)
+    }
+    ### Plot both distributions and quantiles
+    plot(normal_quantiles, dnorm(normal_quantiles, mu, sigma), type = "b",
+         xlim = range(normal_quantiles, sample_quantiles), 
+         ylim = range(dnorm(normal_quantiles, mu, sigma), density_quantiles),
+         col = "red", pch = 0,
+         main = "Densities (lines) and quantiles (symbols)")
+    lines(sample_quantiles, density_quantiles, type = "b")
+    
+    barplot(height = rbind(sample_quantiles, normal_quantiles), beside = TRUE,
+            col = c("black", "red"),
+            main = "Sample quantiles vs Normal quantiles")
+    boxplot(x, rnorm(length(x), mu, sigma), border = c("black", "red"),
+            names = c("Sample", "Corresponding Normal"), 
+            main = "Sample quantiles vs Normal quantiles")
+    title(main = if (input$type_QQ == "Light-tailed"){
+      "Sample quantiles drawn from Uniform(0,1) (black) against corresponding normal quantiles (red)"
+    } else if(input$type_QQ == "Heavy-tailed"){
+      expression(bold(paste("Sample quantiles drawn from t"[2], " (black) against corresponding normal quantiles (red)")))
+    } else if(input$type_QQ == "Normal"){
+      paste0("Sample quantiles drawn from Normal(", round(mu, 1), ",", round(sigma^2, 1), ") (black) against corresponding normal quantiles (red)")
+    } else if(input$type_QQ == "Negatively skewed"){
+      "Sample quantiles drawn from Beta(15,2) (black) against corresponding normal quantiles (red)"
+    } else if(input$type_QQ == "Positively skewed"){
+      "Sample quantiles drawn from Gamma(2) (black) against corresponding normal quantiles (red)"
+    }, outer = TRUE, cex.main = 1.5)
+  })
+  
+  ###############################################################################

```

Comment:
###############################################################################
  # Maximum Likelihood Estimation
  ###############################################################################
  # MLE
  output$plot_MLE <- renderPlot(
    if(input$MLE_type == "bern"){
      x <- rbinom(as.numeric(input$MLE_s),1,as.numeric(input$MLE_p))
      m <- seq(from=0,to=1,length.out = 10000)
      likelihood <- 1
      for(i in 1:length(x)){
        if(x[i] == 1){
          likelihood <- likelihood*m
        }
        else{ 
          likelihood <- likelihood*(1-m)
        }
      }
      plot(x = m, y = likelihood,type = "l",xlab="p",ylim = range(likelihood))
      points(x = m[likelihood==max(likelihood)],y = max(likelihood),col="red",pch = 19)
    }
    else if(input$MLE_type == "geom"){
      x <- rgeom(as.numeric(input$MLE_s),as.numeric(input$MLE_p3))
      m <- seq(from=0,to=1,length.out = 10000)
      likelihood <- 1
      for(i in 1:length(x)){
        likelihood <- likelihood*((1-m)^(x[i]))*m
      }
      plot(x=m,y= likelihood,type = "l",xlab = "p",ylim = range(likelihood))
      points(x = m[likelihood==max(likelihood)],y= max(likelihood),col="red",pch = 19)
    }
    else if(input$MLE_type == "pois"){
      x <- rpois(as.numeric(input$MLE_s),as.numeric(input$MLE_lambda2))
      m <- seq(from=0,to=100,length.out = 10000)
      likelihood <- 1
      for(i in 1:length(x)){
        likelihood <- likelihood*exp(-m)*((m)^(x[i]))/factorial(x[i])
      }
      plot(x=m,y= likelihood,type = "l",xlab = "mu",ylim = range(likelihood))
      points(x = m[likelihood==max(likelihood)],y= max(likelihood),col="red",pch = 19)
    }
    else if(input$MLE_type == "exp"){
      x <- rexp(as.numeric(input$MLE_s),as.numeric(input$MLE_lambda))
      m <- seq(from=0,to=100,length.out = 10000)
      likelihood <- 1
      for(i in 1:length(x)){
        likelihood <- likelihood*m*exp(-m*x[i])
      }
      plot(x=m,y= likelihood,type = "l",xlab = "mu",ylim = range(likelihood))
      points(x = m[likelihood==max(likelihood)],y= max(likelihood),col="red",pch = 19)
    }
    else if(input$MLE_type == "norm"){
      x <- rnorm(as.numeric(input$MLE_s),as.numeric(input$MLE_miu),as.numeric(input$MLE_sigma))
      m <- seq(from=as.numeric(input$miu)-20,to=as.numeric(input$miu)+20,length.out = 100)
      n <- seq(from=1,to=as.numeric(input$sigma)+10,length.out = 100)
      likelihood <- matrix(1,nrow = 100,ncol = 100)
      for(j in 1:length(m)){
        for(t in 1:length(n)){
          for(i in 1:length(x)){
            likelihood[j,t] <- likelihood[j,t]*exp(-((x[i]-m[j])^2)/(2*(n[t]^2)))/(n[t]*sqrt(2*pi))
          }
        }
      }
      persp3D(m,n,likelihood, nticks=5, ticktype="detailed",xlab="mu",ylab="sd",zlab="likelihood" )
      
    }
  )